### PR TITLE
SWS-277 allow grouping metrics by label

### PR DESCRIPTION
**Pull request by jotak**
Also change the prom query on request volume to perform "round(sum(irate(" instead of "rate(" , same as in the istio grafana dashboard.

Sum allows grouping.

### CODE REVIEW

**Comment by jotak** [comment]
@lucasponce or @xeviknal some of you want to review?

File: handlers/services.go [hunk 1]
```diff
@@ -71,13 +71,21 @@ func getServiceMetrics(w http.ResponseWriter, r *http.Request, promClientSupplie
 			return
 		}
 	}
+	var byLabelsIn []string
+	var byLabelsOut []string
+	if lblsin, ok := queryParams["byLabelsIn[]"]; ok && len(lblsin) > 0 {
```
Comment (lucasponce) [inline handlers/services.go]: I guess yes, but I need to ask, the "byLabelsIn[]" the "[]" is intended ?

File: prometheus/metrics.go [hunk 2]
```diff
@@ -129,40 +131,65 @@ func getServiceMetricsAsync(api v1.API, namespace string, servicename string, du
...
 
-func fetchRateRange(api v1.API, metricName string, labels string, bounds v1.Range, rateInterval string, ch chan Metric) {
-	query := fmt.Sprintf("rate(%s%s[%s])", metricName, labels, rateInterval)
+func joinLabels(labels []string) string {
+	str := ""
+	if len(labels) > 0 {
+		sep := ""
+		for _, lbl := range labels {
+			str = str + sep + lbl
+			sep = ","
```
Comment (lucasponce) [inline prometheus/metrics.go]: empty sep at the end is correct/harmless ?

Comment (lucasponce) [reply prometheus/metrics.go]: ah, no, it's the first case used.

File: prometheus/metrics.go [hunk 3]
```diff
@@ -129,40 +131,65 @@ func getServiceMetricsAsync(api v1.API, namespace string, servicename string, du
...
+			"sum(rate(%s_sum%s[%s]))%s / sum(rate(%s_count%s[%s]))%s", metricName, labels, rateInterval, groupingAvg,
+			metricName, labels, rateInterval, groupingAvg)
 		fetchRange(api, query, bounds, avgChan)
 	}()
 
 	// Median
 	go func() {
 		query := fmt.Sprintf(
-			"histogram_quantile(0.5, sum(rate(%s_bucket%s[%s])) by (le))", metricName, labels, rateInterval)
+			"histogram_quantile(0.5, sum(rate(%s_bucket%s[%s])) by (le%s))", metricName, labels, rateInterval, groupingQuantile)
```
Comment (lucasponce) [inline prometheus/metrics.go]: I'm not familiar with syntax of p8s on this one.
"le%s" is the go syntax here ?
(first thought it was a typo)

File: handlers/services.go (hunk 1 above)
Comment (jotak) [reply handlers/services.go]: yes, this is how axios generated my array query params, and I could get them this way in go, working well so I didn't looked further.

File: prometheus/metrics.go (hunk 3 above)
Comment (jotak) [reply prometheus/metrics.go]: `groupingQuantile` is a string built above in this form: `,foo,bar` so it translates in `sum (...) by (le,foo,bar)` which is what I want

Comment (lucasponce) [reply prometheus/metrics.go]: ok, it was hard to follow, as I wasn't sure if the le was intentional.

Comment (lucasponce) [reply prometheus/metrics.go]: Minor, if you can add a comment, perhaps it would clarify to others, just the same thing you put in the PR

Comment (jotak) [reply prometheus/metrics.go]: yeah, these strings builder are not the prettiest thing... I'd wish we had a nice DSL to query prometheus.

Comment (jotak) [reply prometheus/metrics.go]: @lucasponce done I added 2 examples
